Add sidebar

sassypocoyo opened the pull request:
- Implemented the structural changes made on #55;
- Added a Menu button that enables the user to open and close the Sidebar;
- Improved view of the Sidebar on small screens (mobile devices);
- Updated `getNameString()` so that entries with more than 3 authors have the following syntax: `Person1, Person2 and N others` where `N` is the number of following authors. This was done with the intent of making the string more compact to avoid weird formatting, mainly on the sidebar where the space is limited.
- Fixed bugs.

> **Note**
> Had a problem with using `getFullDateString()` on the Location component. It gave me a hydration error so I just opted to not use that function on the sidebar locations list for now.

**Menu button:**

![Screenshot from 2022-09-01 01-18-36](https://user-images.githubusercontent.com/80540164/187806957-8d743ab8-e86c-4159-be70-03d99442b93a.png)

**Sidebar:**

![Screenshot from 2022-09-01 01-19-09](https://user-images.githubusercontent.com/80540164/187806983-4c1c7880-78e1-4cc5-b3b1-fa3e8a607cd3.png)

**Mobile view:**

![Screen Shot 2022-09-01 at 01 20 38](https://user-images.githubusercontent.com/80540164/187807015-92bf661c-8175-44cc-96be-244669f9133e.png)

### Future ideas:

- [x] Implement a transition for the opening and closing of the sidebar
- [x] Create a **Sort By** button that would enable sorting the pins by different variables
- [x] Start development of the Leaderboard feature

nelsonmestevao commented on components/Location/Location.tsx:
```diff
@@ -0,0 +1,37 @@
+        <br></br>
+        {date && <i>{getRelativeTimeString(date)}</i>}
+        <br />
+        <span>{name}</span>
```
I think it would be better here to wrap all this around `p` tags instead of those breaks.

sassypocoyo replied:
You mean like this?
![image](https://user-images.githubusercontent.com/80540164/188016752-cb0a49ca-6ada-4da1-9bd8-e12de011d397.png)
Cause it ends up looking very spaced out:
![image](https://user-images.githubusercontent.com/80540164/188016885-7a8e12a1-062c-4ee0-8e30-8dbd806c092e.png)

nelsonmestevao replied:
Yes, and then you can just customize the padding with css.

You could also use <span> for each and then use something like css-flex on the div container.

I don't think it is a good practice to use br for layout purposes.

sassypocoyo replied:
Got it, solved it with css (`margin: 0`). Also made the city and country text a little bigger.

nelsonmestevao commented on lib/utils.ts:
```diff
@@ -26,5 +26,105 @@ export const getNameString = (authors: string[] | string) => {
+    return authors.slice(0, 2).join(', ') + ' and ' + (size - 2) + ' others';
+  }
+
   return authors.slice(0, size - 1).join(', ') + ' and ' + authors[size - 1];
```
I think it would be better to wrap this one around `if size === 2`.

sassypocoyo replied:
Yes, makes a lot more sense. Done!

nelsonmestevao replied:
I think you forgot this :thinking:

nelsonmestevao commented on pages/index.tsx:
```diff
@@ -2,13 +2,30 @@ import dynamic from 'next/dynamic';
-      <Map pins={PLACES} />
-    </ThemeProvider>
+    <div>
+      <input id="menu_checkbox" type={'checkbox'} onClick={menu} />
+      <label className="menu_label" htmlFor="menu_checkbox" title="Menu">
+        <div></div>
+        <div></div>
+        <div></div>
+      </label>
```
All this should be a `Menu` component :muscle:

sassypocoyo replied:
Got it, I'm just not quite sure why having a `isOpen` state would be useful. I would need to set that state manually according to when the Menu checkbox is checked or unchecked right? Not sure I understood the assignment ahaha. I need to learn more about it.

nelsonmestevao commented on package.json:
```diff
@@ -27,6 +27,8 @@
     "test:format": "prettier --check . --ignore-path .gitignore"
   },
   "dependencies": {
+    "@types/react-transition-group": "^4.4.5",
```
Move this to `devDependencies` :pray:

sassypocoyo replied:
Done @nelsonmestevao , also just committed my last work on the Leaderboard feature, it's still a draft. Have a look and tell me what you think!
Also, I know I used that function `changeVariables()` with a time out for the transitions, but it was necessary for it to work! I tried having them both run at the same time using just one variable called `view` (where one transition would be activated by the `false` state and the other by the `true` state of this variable) but it doesn't work. I also tried to add a `transition-delay` to the CSS but had no luck with that either, but anyhow if you figure a way to avoid it hmu.

P.s. I know the `makeLeaderboard()` function is VERY inefficient but I'll improve it later kk.

sassypocoyo commented:
### Update

Made several improvements to the design, particularly of the Sidebar:
- Implemented a new dynamic bar for each item on the Leaderboard that grows according to the amount of pins;
- Made tweaks to the gradients and colors;
- Updated the Sidebar buttons;
- Imported some fonts from Google Fonts to ensure a consistent look throughout devices.

**Desktop:**

<img src="https://user-images.githubusercontent.com/80540164/189778288-dbad0cf9-a149-4c1a-9bc6-22b10a154b55.png"/>

<img src="https://user-images.githubusercontent.com/80540164/189778338-8bd55153-2ebb-4f82-97da-a1905983533a.png"/>

**Mobile:**

<img src="https://user-images.githubusercontent.com/80540164/189778509-0ff786c9-8074-4632-9803-e8101209ce8f.png" align="left" width="47%" style="float: left;"/>

<img src="https://user-images.githubusercontent.com/80540164/189778515-87df0296-844e-448f-8b10-93131e6feb28.png" align="right" width="47%" style="float: right;"/>

nelsonmestevao commented on package.json:
```diff
@@ -27,6 +27,7 @@
     "test:format": "prettier --check . --ignore-path .gitignore"
   },
   "dependencies": {
+    "classnames": "^2.3.1",
```
Checkout https://www.npmjs.com/package/clx :bulb:

sassypocoyo commented:
### Go to pin functionality!

FINALLY figured out how to do this, we can now click any item on the list to travel to that pin. Opening the pop-up automatically is a similar but more complicated process that I'm not yet sure how to implement and honestly believe is not worth it.

I tried to choose a balanced zoom level so that it looks good on most locations, but there will still be some where the pins are too close together to be shown individually, and will show up as a cluster. I can try to make a dynamic zoom level in the future to fix this.

![image](https://user-images.githubusercontent.com/80540164/190821133-394b6acb-1b1b-404e-8539-75494eb87165.png)

![Screenshot from 2022-09-16 23-24-33](https://user-images.githubusercontent.com/80540164/190821464-c1333839-f815-4f33-b5e3-4ecd059c32bc.png)

https://user-images.githubusercontent.com/80540164/190822184-c1503c6b-b776-40d7-bd7b-53c31abe4fd3.mp4

sassypocoyo commented:
### Updated button design

![Screenshot from 2022-09-17 21-00-39](https://user-images.githubusercontent.com/80540164/190874565-910422ca-bdfa-4cb1-bea7-5a203f16d929.png)

https://user-images.githubusercontent.com/80540164/190874549-020335f8-e4d8-4c38-82ac-a93b1aeb9170.mp4

nelsonmestevao reviewed (changes requested):
The project does not compile at the moment 🔍

Test it with `npm run build` :muscle: 

![image](https://user-images.githubusercontent.com/19409687/190874889-a9ed23d4-c273-4e64-a0f1-cb0fb48e7579.png)

sassypocoyo commented:
> The project does not compile at the moment mag
> 
> Test it with `npm run build` muscle
> 
> ![image](https://user-images.githubusercontent.com/19409687/190874889-a9ed23d4-c273-4e64-a0f1-cb0fb48e7579.png)

Fixed ;)

sassypocoyo commented:
The sidebar now dynamically closes on mobile devices after clicking a location, this allows the user to see the map move to the selected location. Since the sidebar occupies all the screen on mobile devices, it blocked the map view before.

sassypocoyo commented:
@nelsonmestevao

Since university starts soon, I'm not going to be as active on CeSIUMpeloMundo as I have been. Considering that the Sidebar already presents a substantial amount of functionality to the user (and looks cool af xD), I would love to finally merge this PR, after resolving any problems that you may find when reviewing of course. Do you agree?

Thank you for your help! :muscle:

RuiL1904 commented on lib/utils.ts:
```diff
@@ -26,5 +26,55 @@ export const getNameString = (authors: string[] | string) => {
+
+  let dlon = lon2 - lon1;
+  let dlat = lat2 - lat1;
+  let a =
+    Math.pow(Math.sin(dlat / 2), 2) +
+    Math.cos(lat1) * Math.cos(lat2) * Math.pow(Math.sin(dlon / 2), 2);
+
+  let c = 2 * Math.asin(Math.sqrt(a));
+
+  let r = 6371;
+
+  return c * r;
+}
+
+export const sortByDistance = (a, b) => {
+  const x0 = 41.56157392223945;
+  const y0 = -8.397397824887639;
+
+  const x1 = a.coordinates[0];
+  const y1 = a.coordinates[1];
+  const x2 = b.coordinates[0];
+  const y2 = b.coordinates[1];
+
+  const d1 = distance(x0, x1, y0, y1);
+  const d2 = distance(x0, x2, y0, y2);
+
+  return d1 - d2;
+};
+
+export const sortByOldest = (a, b) => {
+  return (
+    DateTime.fromISO(a.date).toMillis() - DateTime.fromISO(b.date).toMillis()
+  );
+};
+
+export const sortByLatest = (a, b) => {
+  return (
+    DateTime.fromISO(b.date).toMillis() - DateTime.fromISO(a.date).toMillis()
+  );
 };
```
Don't you think it would make more sense to have this functions on `~/components/Sidebar/utils.ts` or even `~/components/Leaderboard/utils.ts`?

sassypocoyo replied:
Done!
Moved the functions to their respective `utils.ts` document. However, the ones that are used in more than one component (`getRelativeTimeString`, `getNameString` and `sortByOldest`) stayed on the same place.

RuiL1904 reviewed (commented):
This happens when on build mode. Can you fix it? 🙏🏼

![image](https://user-images.githubusercontent.com/76881129/192101970-ecd601a2-6d04-4b1d-86d5-4d040fcc30d7.png)

sassypocoyo commented:
> This happens when on build mode. Can you fix it? 🙏🏼
> 
> ![image](https://user-images.githubusercontent.com/76881129/192101970-ecd601a2-6d04-4b1d-86d5-4d040fcc30d7.png)

Think it's working normally now :+1:
